### PR TITLE
[#2219] feat: Introduce ShuffleBlockIdManagerFactory and PartitionedShuffleBlockIdManager

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes were proposed in this pull request?

- Introduce a ShuffleBlockIdManagerFactory to create configured implementation.
- Introduce PartitionedShuffleBlockIdManager to individual bitmap for each partition.

### Why are the changes needed?

Fix: #2219

image

This is the maximum partition num with node metrics in our cluster, it is ~410K, it cost heap size 410K * 50KiB = 2GiB if we say a bitmap related a partition avg 10000 blocks(cost 50KiB heap).

I think this is worth to choose this new policy in our production cluster.

### Does this PR introduce _any_ user-facing change?

- Config option: rss.server.blockIdStrategyClass

### How was this patch tested?

Test Locally, check result by `arthas`

- Common case, test with default config
```
[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[0].shuffleBlockIdManager'
@DefaultShuffleBlockIdManager[
    LOG=@Log4jLogger[org.apache.logging.slf4j.Log4jLogger@3f211bae],
    partitionsToBlockIds=@ConcurrentHashMapForJDK8[isEmpty=false;size=1],
]

[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[0].shuffleBlockIdManager.partitionsToBlockIds.get("application_1729845342052_431031_1730297736664") '
@ConcurrentHashMapForJDK8[
    @Integer[0]:@Roaring64NavigableMap[][isEmpty=false;size=10],
    @Integer[1]:@Roaring64NavigableMap[][isEmpty=false;size=10],
    @Integer[2]:@Roaring64NavigableMap[][isEmpty=false;size=10],
]

```
- Config to PartitionedShuffleBlockIdManager
```
[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[0].shuffleBlockIdManager'
@PartitionedShuffleBlockIdManager[
    LOG=@Log4jLogger[org.apache.logging.slf4j.Log4jLogger@6aa6fea0],
    partitionsToBlockIds=@ConcurrentHashMap[isEmpty=false;size=1],
]

[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[0].shuffleBlockIdManager.partitionsToBlockIds'
@ConcurrentHashMap[
    @String[application_1729845342052_432324_1730299397916]:@ConcurrentHashMapForJDK8[isEmpty=false;size=2],
]
[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[0].shuffleBlockIdManager.partitionsToBlockIds.get("application_1729845342052_432324_1730299397916")'
@ConcurrentHashMapForJDK8[
    @Integer[0]:@ConcurrentHashMap[isEmpty=false;size=2000],
    @Integer[1]:@ConcurrentHashMap[isEmpty=false;size=2000],
]
```
- Use client app level config
submit two app with `spark.rss.client.blockIdStrategyClass=org.apache.uniffle.server.block.DefaultShuffleBlockIdManager` and `spark.rss.client.blockIdStrategyClass=org.apache.uniffle.server.block.PartitionedShuffleBlockIdManager`

```
[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[0].shuffleBlockIdManager'
@DefaultShuffleBlockIdManager[
    LOG=@Log4jLogger[org.apache.logging.slf4j.Log4jLogger@ed3b52a],
    partitionsToBlockIds=@ConcurrentHashMapForJDK8[isEmpty=false;size=1],
]
[arthas@264]$ vmtool  --action getInstances --className org.apache.uniffle.server.ShuffleTaskInfo --express 'instances[1].shuffleBlockIdManager'
@PartitionedShuffleBlockIdManager[
    LOG=@Log4jLogger[org.apache.logging.slf4j.Log4jLogger@5fc14628],
    partitionsToBlockIds=@ConcurrentHashMap[isEmpty=false;size=1],
]
```

Log example
```
[2024-10-30 23:01:37.585] [Grpc-709] [INFO] ShuffleTaskInfo - application_1729845342052_432387_1730300440568 use app configured ShuffleBlockIdManager to org.apache.uniffle.server.block.DefaultShuffleBlockIdManager@3c6b75a8
```

### CODE REVIEW

**Comment by jerqi** [comment]
For some little shuflfe, `PartionedShuffleBlock` may have worse effect. Like mixed load, there are many little shuffle.

**Comment by jerqi** [comment]
> > Should it become an application level config option?
> 
> @jerqi Sorry for the late reply.
> 
> It could be better if it become an app level request option. But It could be a little complex as follow
> 
> * Config this option in client conf. (Just list to here, this is easy to do)
> * Make ShuffleBlockIdManager as a member of ShuffleTaskInfo which created follow the config from registerShuffleRequest
> 
> Could you explain the benefit?

We can have a ShuffleBlockIdManager, but ShuffleTaskInfo has its BlockId strategy.

**Comment by maobaolong** [comment]
@jerqi Thanks for your suggestion, how about this commit? https://github.com/apache/incubator-uniffle/pull/2227/commits/93564afbc9127d44ebedd773a9401176f4f9a2aa

**Comment by maobaolong** [comment]
@jerqi Update the name of config key, PTAL

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 1]
```diff
@@ -719,6 +720,11 @@ public class ShuffleServerConf extends RssBaseConf {
           .booleanType()
           .defaultValue(false)
           .withDescription("Whether to enable app detail log");
+  public static final ConfigOption<String> SERVER_BLOCK_ID_STRATEGY_CLASS =
+      ConfigOptions.key("rss.server.blockIdStrategyClass")
+          .stringType()
+          .defaultValue(DefaultShuffleBlockIdManager.class.getName())
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:726]: DefaultShuffleBlockStrategy?

**Comment by EnricoMi** [comment]
In code, there is the term "manager", while in (user-facing) config there is the term "strategy". I think there should be one term. I'd prefer "manager".

**Comment by maobaolong** [comment]
@EnricoMi Thanks for your suggestion, in fact this is my original `term` I used for this PR, I'm open to change to `strategy` or revert to `manager`.

**Comment by jerqi** [comment]
> In code, there is the term "manager", while in (user-facing) config there is the term "strategy". I think there should be one term. I'd prefer "manager".

It's better to have one term. My concern about `manager` is that every `TaskInfo` will have a manager. In my view, we would better not have the same managers in a server. WDYT?

**Comment by EnricoMi** [comment]
I don't see a problem having different **types** of manager in the server. The name `ShuffleTaskManager` clearly says this manager is about shuffle tasks, whereas the `ShuffleBlockIdManager` manages shuffle block ids.

I'd prefer "Manager" over "Strategy".

**Comment by maobaolong** [comment]
@jerqi @EnricoMi Thanks for your discussion and suggestion, I updated the PR, PTAL.

File: docs/server_guide.md [hunk 2]
```diff
@@ -121,6 +121,7 @@ This document will introduce how to deploy Uniffle shuffle servers.
...
+| rss.server.blockIdManagerClass                           | org.apache.uniffle.server.block.DefaultShuffleBlockIdManager           | The block id manager class. It is used to manage block id.                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
```
Comment (jerqi) [inline docs/server_guide.md:124]: Could you add more description about defaultShuffleBlockIdManger?

File: server/src/main/java/org/apache/uniffle/server/block/DefaultShuffleBlockIdManager.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.common.exception.InvalidRequestException;
+import org.apache.uniffle.common.exception.RssException;
+import org.apache.uniffle.common.util.BlockIdLayout;
+import org.apache.uniffle.common.util.JavaUtils;
+import org.apache.uniffle.common.util.RssUtils;
+import org.apache.uniffle.server.ShuffleTaskInfo;
+
+public class DefaultShuffleBlockIdManager implements ShuffleBlockIdManager {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/block/DefaultShuffleBlockIdManager.java:41]: Could you add some comments about the behaviour of this class?

File: server/src/main/java/org/apache/uniffle/server/block/PartitionedShuffleBlockIdManager.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.common.exception.RssException;
+import org.apache.uniffle.common.util.BlockIdLayout;
+import org.apache.uniffle.common.util.JavaUtils;
+import org.apache.uniffle.common.util.RssUtils;
+import org.apache.uniffle.server.ShuffleTaskInfo;
+
+public class PartitionedShuffleBlockIdManager implements ShuffleBlockIdManager {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/block/PartitionedShuffleBlockIdManager.java:37]: Could you add more comment about this class?

File: docs/server_guide.md (hunk 2 above)
Comment (maobaolong) [reply docs/server_guide.md:124]: done

File: server/src/main/java/org/apache/uniffle/server/block/DefaultShuffleBlockIdManager.java (hunk 3 above)
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/block/DefaultShuffleBlockIdManager.java:44]: done

File: server/src/main/java/org/apache/uniffle/server/block/PartitionedShuffleBlockIdManager.java (hunk 4 above)
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/block/PartitionedShuffleBlockIdManager.java:40]: done

**Comment by maobaolong** [comment]
@jerqi Thanks for your comment, updated the description and docs, PTAL.

**Review by jerqi — APPROVED** [review]

**Comment by maobaolong** [comment]
@jerqi Thanks for your review. merge it.

**Pull request merged**